bytecomp/matching.ml

```diff
@@ -2874,81 +3337,108 @@ let is_mutable p = find_in_pat have_mutable_field p
+        in
+        check_total total action.expr raise_num handler_fun
       with
       | Unused -> assert false (* ; handler_fun() *)
       end
   | Total ->
       let pm =
-        { cases = List.map (fun (pat, act) -> ([pat], act)) pat_act_list;
-          args = [arg, Strict] ;
-          default = []} in
-      let (lambda, total) = compile_match repr partial (start_ctx 1) pm in
-      assert (jumps_is_empty total) ;
-      lambda
-
+        { cases;
+          args = [arg, Strict, loc];
+          default = [];
+        }
+      in
+      let action, total =
+        compile_match loc repr partial (start_ctx loc 1) pm
+      in
+      assert (jumps_is_empty total);
+      action.expr
 
-let partial_function loc () =
+let partial_function loc () : Lambda.block =
   let slot =
     transl_extension_path loc
       Env.initial_safe_string Predef.path_match_failure
   in
-  let (fname, line, char) = Location.get_pos_info loc.Location.loc_start in
-  Lprim(Praise Raise_regular, [Lprim(Pmakeblock(0, Immutable, None),
-          [slot; Lconst(Const_block(0,
-                   [Const_base(Const_string (fname, None));
-                    Const_base(Const_int line);
-                    Const_base(Const_int char)]))], loc)], loc)
+  let fname, line, char = Location.get_pos_info loc.Location.loc_start in
+  let expr =
+    Lprim(Praise (Raise_regular None), [Lprim(Pmakeblock(0, Immutable, None),
+            [slot; Lconst(Const_block(0,
+                     [Const_base(Const_string (fname, None));
+                      Const_base(Const_int line);
+                      Const_base(Const_int char)]), loc)], loc)], loc)
+  in
+  Lambda.block loc expr
 
-let for_function loc repr param pat_act_list partial =
-  compile_matching repr (partial_function loc) param pat_act_list partial
+let for_function loc repr param (cases : pattern cases) partial =
+  compile_matching loc repr (partial_function loc) param cases partial
 
 (* In the following two cases, exhaustiveness info is not available! *)
-let for_trywith param pat_act_list =
-  compile_matching None
-    (fun () -> Lprim(Praise Raise_reraise, [param], Location.none))
-    param pat_act_list Partial
+let for_trywith param (cases : pattern cases) =
+  let loc =
+    (* The location is supposed to point just after the "with". *)
+    match cases with
+    | (_pat, block)::_ -> block.block_loc
+    | [] -> Location.none
+  in
+  let handler =
+    compile_matching loc None
+      (fun () ->
+        let expr =
+          (* The [Location.none] ensures that the try-with itself does not
+             appear as a "reraise" frame in any backtrace. *)
+          Lprim(Praise (Raise_reraise (Some Location.none)), [param], loc)
+        in
+        Lambda.block loc expr)
+      param cases Partial
+  in
+  Lambda.block loc handler
 
-let simple_for_let loc param pat body =
-  compile_matching None (partial_function loc) param [pat, body] Partial
+let simple_for_let loc param pat body body_loc =
```

Comment:
Why is this function not taking a `(body : Lambda.block)` directly?

Reply:
Fixed